examples/ex28p.cpp

```diff
@@ -0,0 +1,522 @@
+      {
+         pmesh->UniformRefinement();
+      }
+   }
+
+   // 6. Define a parallel finite element space on the parallel mesh. Here we
+   //    use vector finite elements, i.e. dim copies of a scalar finite element
+   //    space. We use the ordering by vector dimension (the last argument of
+   //    the FiniteElementSpace constructor) which is expected in the systems
+   //    version of BoomerAMG preconditioner. For NURBS meshes, we use the
+   //    (degree elevated) NURBS space associated with the mesh nodes.
+   FiniteElementCollection *fec;
+   ParFiniteElementSpace *fespace;
+   const bool use_nodal_fespace = pmesh->NURBSext && !amg_elast;
+   if (use_nodal_fespace)
+   {
+      fec = NULL;
+      fespace = (ParFiniteElementSpace *)pmesh->GetNodes()->FESpace();
+   }
+   else
+   {
+      fec = new H1_FECollection(order, dim);
+      if (reorder_space)
+      {
+         fespace = new ParFiniteElementSpace(pmesh, fec, dim, Ordering::byNODES);
+      }
+      else
+      {
+         fespace = new ParFiniteElementSpace(pmesh, fec, dim, Ordering::byVDIM);
+      }
+   }
+   HYPRE_Int size = fespace->GlobalTrueVSize();
+   if (myid == 0)
+   {
+      cout << "Number of finite element unknowns: " << size << endl
+           << "Assembling: " << flush;
+   }
+
+   // 7. Determine the list of true (i.e. parallel conforming) essential
+   //    boundary dofs. In this example, there are no essential boundary
+   //    conditions in the usual sense, but we leave the machinery here for
+   //    users to modify if they wish.
+   Array<int> ess_tdof_list, ess_bdr(pmesh->bdr_attributes.Max());
+   ess_bdr = 0;
+   fespace->GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
+
+   // 8. Set up the parallel linear form b(.) which corresponds to the
+   //    right-hand side of the FEM linear system. In this case, b_i equals the
+   //    boundary integral of f*phi_i where f represents a "pull down" force on
+   //    the Neumann part of the boundary and phi_i are the basis functions in
+   //    the finite element fespace. The force is defined by the object f, which
+   //    is a vector of Coefficient objects. The fact that f is non-zero on
+   //    boundary attribute 2 is indicated by the use of piece-wise constants
+   //    coefficient for its last component.
+   VectorArrayCoefficient f(dim);
+   for (int i = 0; i < dim-1; i++)
+   {
+      f.Set(i, new ConstantCoefficient(0.0));
+   }
+
+   // 9. Put a leftward force on the right side of the trapezoid
+   {
+      Vector push_force(pmesh->bdr_attributes.Max());
+      push_force = 0.0;
+      push_force(1) = -5.0e-2; // index 1 attribute 2
+      f.Set(0, new PWConstCoefficient(push_force));
+   }
+
+   // 10. Set up constraint matrix to constrain normal displacement (but
+   //     allow tangential displacement) on specified boundaries.
+   Array<int> constraint_atts(2);
+   constraint_atts[0] = 1;  // attribute 1 bottom
+   constraint_atts[1] = 4;  // attribute 4 left side
```

Comment:
Can this be moved down to where the `BuildNormalConstraints` function is actually called?